src/DryIoc/Container.cs

```diff
@@ -11513,21 +11532,53 @@ internal object TryGetOrAddViaFactoryDelegate(int id, FactoryDelegate createValu
                 // Otherwise our own item was added and we can skip this step.
                 if (newMap == oldMap)
                 {
-                    itemRef = map.GetEntryOrDefault(id);
-                    // if someone added the item already there is a chance that it has a created value, so let's check it
-                    if (itemRef.Value != NoItem)
-                        return itemRef.Value;
+                    itemRef = map.GetSurePresentEntry(id);
+                    
+                    // If someone added the item already there is a chance it has a created value, so let's check it.
+                    // Otherwise wait until the other thread creates populates the entry.
+                    if (itemRef.Value == NoItem) 
+                    {
+#if SUPPORTS_SPIN_WAIT
+                        var spinWait = new SpinWait();
+                        while (itemRef.Value == NoItem)
+                            spinWait.SpinOnce();
+#else
+                    lock (itemRef) 
+                        while (itemRef.Value == NoItem)
+                            Monitor.Wait(itemRef);
+#endif
+                    }
+
+                    return itemRef.Value;
                 }
             }
             else 
-                itemRef = Ref.SwapAndGetNewValue(ref map, itemRef, (x, i) => x.AddOrKeepEntry(i)).GetEntryOrDefault(id);
+            {
+                var otherItemRef = Ref.SwapAndGetNewValue(ref map, itemRef, (x, i) => x.AddOrKeepEntry(i)).GetSurePresentEntry(id);
+                if (otherItemRef != itemRef) 
+                {
+#if SUPPORTS_SPIN_WAIT
+                    var spinWait = new SpinWait();
+                    while (otherItemRef.Value == NoItem)
+                        spinWait.SpinOnce();
```

Comment:
what if `createValue` failed? i mean there were 2 calls. first failed, than second will spin until third call will be done?

Reply:
If it failed then the exception will stop everything anyway.

Reply:
one thread said it will create value and factory delegate failed, the other thread was spinning, how that other will pass? lock, if locked delegate failed, allowed others to proceed.

Reply:
The app will crash with all the threads, so ot does not matter if one was spinned or not.

Reply:
I am not sure what do you mean? 

You mean the spinned thread will leak outside of the app? How?

Reply:
what if do try catch of one thread,  any chance second thread will loop forever?